File: ai-services/app/monitoring.py
"""
Monitoring and Logging Utilities for AI Services
Provides structured logging and error tracking
"""
import logging
import json
import time
from typing import Any, Dict, Optional
from functools import wraps
from datetime import datetime

# Configure structured logging
logger = logging.getLogger(__name__)


class StructuredLogger:
    """Structured logging for Better Stack integration"""
    
    @staticmethod
    def log(
        level: str,
        message: str,
        context: Optional[Dict[str, Any]] = None
    ):
        """Log structured message"""
        log_entry = {
            "level": level,
            "message": message,
            "timestamp": datetime.utcnow().isoformat(),
            "service": "ai-services",
            **(context or {})
        }
        
        log_method = getattr(logger, level.lower(), logger.info)
        log_method(json.dumps(log_entry))
    
    @staticmethod
    def info(message: str, context: Optional[Dict[str, Any]] = None):
        StructuredLogger.log("info", message, context)
    
    @staticmethod
    def warning(message: str, context: Optional[Dict[str, Any]] = None):
        StructuredLogger.log("warning", message, context)
    
    @staticmethod
    def error(message: str, context: Optional[Dict[str, Any]] = None):
        StructuredLogger.log("error", message, context)
    
    @staticmethod
    def debug(message: str, context: Optional[Dict[str, Any]] = None):
        StructuredLogger.log("debug", message, context)


class PerformanceMonitor:
    """Monitor operation performance"""
    
    def __init__(self, operation: str, context: Optional[Dict[str, Any]] = None):
        self.operation = operation
        self.context = context or {}
        self.start_time = time.time()
        
        StructuredLogger.debug(f"Starting: {operation}", self.context)
    
    def finish(self, additional_context: Optional[Dict[str, Any]] = None):
        """Finish monitoring and log duration"""
        duration_ms = (time.time() - self.start_time) * 1000
        context = {
            **self.context,
            **(additional_context or {}),
            "duration_ms": duration_ms
        }
        
        StructuredLogger.info(f"Completed: {self.operation}", context)
        
        # Warn on slow operations
        if duration_ms > 5000:
            StructuredLogger.warning(f"Slow operation: {self.operation}", context)
        
        return duration_ms
    
    def error(self, error: Exception, additional_context: Optional[Dict[str, Any]] = None):
        """Log error with duration"""
        duration_ms = (time.time() - self.start_time) * 1000
        context = {
            **self.context,
            **(additional_context or {}),
            "duration_ms": duration_ms,
            "error": str(error),
            "error_type": type(error).__name__
        }
        
        StructuredLogger.error(f"Failed: {self.operation}", context)


def monitor_performance(operation_name: str):
    """Decorator to monitor function performance"""
    def decorator(func):
        @wraps(func)
        async def async_wrapper(*args, **kwargs):
            monitor = PerformanceMonitor(operation_name, {
                "function": func.__name__
            })
            try:
                result = await func(*args, **kwargs)
                monitor.finish()
                return result
            except Exception as e:
                monitor.error(e)
                raise
        
        @wraps(func)
        def sync_wrapper(*args, **kwargs):
            monitor = PerformanceMonitor(operation_name, {
                "function": func.__name__
            })
            try:
                result = func(*args, **kwargs)
                monitor.finish()
                return result
            except Exception as e:
                monitor.error(e)
                raise
        
        # Return appropriate wrapper based on function type
        import asyncio
        if asyncio.iscoroutinefunction(func):
            return async_wrapper
        return sync_wrapper
    
    return decorator


def track_api_call(endpoint: str, method: str, status_code: int, duration_ms: float):
    """Track API call metrics"""
    level = "error" if status_code >= 500 else "warning" if status_code >= 400 else "info"
    
    StructuredLogger.log(level, f"API Call: {method} {endpoint}", {
        "endpoint": endpoint,
        "method": method,
        "status_code": status_code,
        "duration_ms": duration_ms
    })


def track_llm_call(
    model: str,
    prompt_tokens: int,
    completion_tokens: int,
    duration_ms: float,
    success: bool = True
):
    """Track LLM API call metrics"""
    StructuredLogger.info("LLM Call", {
        "model": model,
        "prompt_tokens": prompt_tokens,
        "completion_tokens": completion_tokens,
        "total_tokens": prompt_tokens + completion_tokens,
        "duration_ms": duration_ms,
        "success": success
    })


def track_pdf_processing(
    note_id: str,
    operation: str,
    pages: int,
    duration_ms: float,
    success: bool = True
):
    """Track PDF processing metrics"""
    StructuredLogger.info("PDF Processing", {
        "note_id": note_id,
        "operation": operation,
        "pages": pages,
        "duration_ms": duration_ms,
        "success": success
    })


def report_error(error: Exception, context: Optional[Dict[str, Any]] = None):
    """Report error with context"""
    StructuredLogger.error(str(error), {
        **(context or {}),
        "error_type": type(error).__name__,
        "error_message": str(error)
    })


# Export logger instance
log = StructuredLogger()

File: src/lib/monitoring.ts
/**
 * Monitoring and Logging Utilities
 * Provides structured logging and monitoring helpers
 */

import * as Sentry from "@sentry/react";

export type LogLevel = "debug" | "info" | "warning" | "error" | "critical";

interface LogContext {
  userId?: string;
  action?: string;
  component?: string;
  duration?: number;
  [key: string]: any;
}

/**
 * Structured logging function
 * Logs to console in development and sends to Better Stack in production
 */
export const log = (
  level: LogLevel,
  message: string,
  context?: LogContext
) => {
  const timestamp = new Date().toISOString();
  const logEntry = {
    level,
    message,
    timestamp,
    environment: import.meta.env.MODE,
    ...context,
  };

  // Console logging
  const consoleMethod = level === "error" || level === "critical" ? "error" : 
                       level === "warning" ? "warn" : 
                       level === "debug" ? "debug" : "log";
  
  if (import.meta.env.DEV) {
    console[consoleMethod](`[${level.toUpperCase()}]`, message, context);
  } else {
    // Structured JSON logging for Better Stack
    console[consoleMethod](JSON.stringify(logEntry));
  }

  // Add breadcrumb to Sentry
  if (level === "error" || level === "critical") {
    Sentry.addBreadcrumb({
      message,
      level: level === "critical" ? "fatal" : level,
      data: context,
    });
  }
};

/**
 * Performance monitoring helper
 * Tracks operation duration and logs performance metrics
 */
export class PerformanceMonitor {
  private startTime: number;
  private operation: string;
  private context?: LogContext;

  constructor(operation: string, context?: LogContext) {
    this.operation = operation;
    this.context = context;
    this.startTime = performance.now();
    
    log("debug", `Starting: ${operation}`, context);
  }

  finish(additionalContext?: LogContext) {
    const duration = performance.now() - this.startTime;
    const context = { ...this.context, ...additionalContext, duration };
    
    log("info", `Completed: ${this.operation}`, context);
    
    // Send to Sentry as custom metric
    Sentry.metrics.distribution(
      `operation.duration.${this.operation.replace(/\s+/g, "_").toLowerCase()}`,
      duration,
      {
        unit: "millisecond",
        tags: this.context,
      }
    );
    
    return duration;
  }

  error(error: Error, additionalContext?: LogContext) {
    const duration = performance.now() - this.startTime;
    const context = { ...this.context, ...additionalContext, duration };
    
    log("error", `Failed: ${this.operation}`, context);
    Sentry.captureException(error, { extra: context });
  }
}

/**
 * Track custom metrics
 */
export const trackMetric = (
  name: string,
  value: number,
  unit: string = "none",
  tags?: Record<string, string>
) => {
  log("info", `Metric: ${name}`, { value, unit, ...tags });
  
  Sentry.metrics.gauge(name, value, {
    unit,
    tags,
  });
};

/**
 * Track user actions
 */
export const trackAction = (
  action: string,
  category: string,
  context?: LogContext
) => {
  log("info", `Action: ${action}`, { category, ...context });
  
  Sentry.addBreadcrumb({
    message: action,
    category,
    level: "info",
    data: context,
  });
  
  // Increment action counter
  Sentry.metrics.increment(`action.${category}.${action.replace(/\s+/g, "_").toLowerCase()}`, 1, {
    tags: context as Record<string, string>,
  });
};

/**
 * Track API calls
 */
export const trackApiCall = (
  endpoint: string,
  method: string,
  statusCode: number,
  duration: number,
  context?: LogContext
) => {
  const level = statusCode >= 500 ? "error" : statusCode >= 400 ? "warning" : "info";
  
  log(level, `API Call: ${method} ${endpoint}`, {
    statusCode,
    duration,
    ...context,
  });
  
  Sentry.metrics.distribution("api.response_time", duration, {
    unit: "millisecond",
    tags: {
      endpoint,
      method,
      status: statusCode.toString(),
    },
  });
};

/**
 * Track page views
 */
export const trackPageView = (pageName: string, context?: LogContext) => {
  log("info", `Page View: ${pageName}`, context);
  
  Sentry.addBreadcrumb({
    message: `Viewed ${pageName}`,
    category: "navigation",
    level: "info",
    data: context,
  });
  
  Sentry.metrics.increment("page.view", 1, {
    tags: { page: pageName },
  });
};

/**
 * Monitor React Query operations
 */
export const monitorQuery = (
  queryKey: string,
  status: "loading" | "success" | "error",
  duration?: number,
  error?: Error
) => {
  if (status === "error" && error) {
    log("error", `Query failed: ${queryKey}`, { duration });
    Sentry.captureException(error, {
      extra: { queryKey, duration },
    });
  } else if (status === "success" && duration) {
    log("debug", `Query succeeded: ${queryKey}`, { duration });
    
    if (duration > 1000) {
      log("warning", `Slow query: ${queryKey}`, { duration });
    }
  }
};

/**
 * Error reporting with context
 */
export const reportError = (
  error: Error,
  context?: LogContext,
  level: "error" | "critical" = "error"
) => {
  log(level, error.message, { ...context, stack: error.stack });
  
  Sentry.captureException(error, {
    level: level === "critical" ? "fatal" : "error",
    extra: context,
  });
};

/**
 * Check if monitoring is enabled
 */
export const isMonitoringEnabled = (): boolean => {
  return !!import.meta.env.VITE_SENTRY_DSN;
};

/**
 * Get monitoring status
 */
export const getMonitoringStatus = () => {
  return {
    sentryEnabled: isMonitoringEnabled(),
    environment: import.meta.env.MODE,
    analyticsEnabled: import.meta.env.VITE_ENABLE_ANALYTICS === "true",
  };
};
